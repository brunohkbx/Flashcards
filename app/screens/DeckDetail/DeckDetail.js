import React, { Component } from 'react';
import { connect } from 'react-redux';
import { View, FlatList, KeyboardAvoidingView } from 'react-native';
import { TextInput } from 'react-native-paper';
import ScreenToolbar from './ScreenToolbar';
import Flashcard from '../../components/Flashcard';
import { FlashcardIcon } from '../../components/Icons';
import NoContent from './NoContent';
import { Header } from 'react-navigation';

export class DeckDetail extends Component {
  static navigationOptions = ({ navigation }) => {
    return {
      header: <ScreenToolbar
                navigation={navigation}
                onDelete={navigation.getParam('openDialog')}
              />
    };
  };

  renderIcon = ({size, color}) => <FlashcardIcon size={size} color={color} />

  renderItem = ({ item }) => {
    return (
      <Flashcard>
        <TextInput
          label="Question"
          value={item.question}
          multiline
          editable={false}
        />
        <TextInput
          label="Answer"
          value={item.answer}
          multiline
          editable={false}
        />
      </Flashcard>
    );
  }

  componentDidMount() {
    const { navigation } = this.props;

    navigation.setParams({ openDialog: this.openDialog });
  }

  render() {
    const { deck } = this.props;

    return (
      <View style={{ flex: 1}}>
        <KeyboardAvoidingView
          style={{flex: 1}}
          behavior="padding"
          enabled
          keyboardVerticalOffset={Header.HEIGHT + 30}
        >
        {deck.questions.length > 0 ?
          <FlatList
            data={deck.questions}
            renderItem={this.renderItem}
            keyExtractor={item => item.id}
          />
          :
          <NoContent/>
        }
        </KeyboardAvoidingView>
      </View>
    );
  }
}

const mapStateToProps = ({ decks }, { navigation }) => {
  const { deckId } = navigation.state.params;

  return {
    deck: decks[deckId]
  };
};

export default connect(mapStateToProps)(DeckDetail);
