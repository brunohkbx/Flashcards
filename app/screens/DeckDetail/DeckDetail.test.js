import React from 'react';
import { shallow } from 'enzyme';
import { DeckDetail } from './DeckDetail';

describe('DeckDetail', () => {
  const setup = propOverrides => {
    const defaultProps = Object.assign({
      navigation: {
        setParams: jest.fn()
      },
      deck: { questions: [] }
    }, propOverrides);

    const wrapper = shallow(<DeckDetail {...defaultProps} />);
    const wrapperInstance = wrapper.instance();

    const pressOnEditButton = () => {
      wrapper
        .find('NoContent')
        .dive()
        .find('withTheme(Button)')
        .simulate('press');
    };

    return { wrapper, wrapperInstance, pressOnEditButton };
  };

  it('renders properly when deck has questions', () => {
    const { wrapper } = setup({
      deck: {
        questions: [
          {
            question: 'What is React?',
            answer: 'A library for managing user interfaces'
          }
        ]
      }
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('renders properly when deck does not have questions', () => {
    const { wrapper } = setup({ deck: { questions: [] }});

    expect(wrapper).toMatchSnapshot();
  });

  describe('renderItem', () => {
    it('renders properly', () => {
      const { wrapperInstance } = setup();

      expect(
        wrapperInstance.renderItem(
          {
            item: {
              question: 'What is React?',
              answer: 'A library for managing user interfaces'
            }
          }
        )
      ).toMatchSnapshot();
    });
  });

  describe('componentDidMount', () => {
    it('sets openDialog and quizDisabled as navigation params', () => {
      const mockNavigation = { setParams: jest.fn() };
      const { wrapperInstance } = setup({ navigation: mockNavigation });

      expect(mockNavigation.setParams).toHaveBeenCalledWith(
        {
          openDialog: wrapperInstance.openDialog,
          quizDisabled: wrapperInstance.props.deck.questions.length === 0
        }
      );
    });
  });

  test('EditButton of NoContent screen navigates to Edit screen', () => {
    const mockNavigation = { navigate: jest.fn(), setParams: jest.fn() };
    const { pressOnEditButton } = setup(
      {
        deck: { id: '1', questions: [] },
        navigation: mockNavigation
      }
    );

    pressOnEditButton();

    expect(mockNavigation.navigate).toHaveBeenCalledWith(
      'EditDeck',
      { deckId: '1' }
    );
  });
});
