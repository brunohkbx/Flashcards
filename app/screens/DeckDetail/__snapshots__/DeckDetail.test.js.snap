// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeckDetail renderIcon renders properly 1`] = `
<FlashcardIcon
  color="#ffffff"
  size={24}
/>
`;

exports[`DeckDetail renderItem renders properly 1`] = `
<Flashcard>
  <withTheme(TextInput)
    editable={false}
    label="Question"
    multiline={true}
    value="What is React?"
  />
  <withTheme(TextInput)
    editable={false}
    label="Answer"
    multiline={true}
    value="A library for managing user interfaces"
  />
</Flashcard>
`;

exports[`DeckDetail renders properly when deck does not have questions 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <KeyboardAvoidingView
    behavior="padding"
    enabled={true}
    keyboardVerticalOffset={94}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <NoContent />
  </KeyboardAvoidingView>
</Component>
`;

exports[`DeckDetail renders properly when deck has questions 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <KeyboardAvoidingView
    behavior="padding"
    enabled={true}
    keyboardVerticalOffset={94}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <FlatList
      data={
        Array [
          Object {
            "answer": "A library for managing user interfaces",
            "question": "What is React?",
          },
        ]
      }
      disableVirtualization={false}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onEndReachedThreshold={2}
      renderItem={[Function]}
      scrollEventThrottle={50}
      updateCellsBatchingPeriod={50}
      windowSize={21}
    />
  </KeyboardAvoidingView>
</Component>
`;
